### DISCUSSION TITLE
Injecting actorRef into the machine's system for the purpose of testing

### DISCUSSION THREAD

**Discussion by bhvngt** [statelyai/xstate#4135] [Q&A] [answered]
I am using xstate `v5`. I was tinkering with the [receptionist](https://stately.ai/blog/announcing-xstate-v5-beta#actor-system) pattern and was looking at injecting `actorRef` into machine's system for the purpose of testing. Since, I believe, systemRef are different from actorRef injected through the [provider api](https://stately.ai/docs/xstate-v5/machines#provide-implementations), I was wondering if it is safe to use [_set](https://github.com/statelyai/xstate/blob/be59508116f6c746f30375055047094102b28002/packages/core/src/system.ts#L27) method to inject external actorRef?

**Comment by davidkpiano** [comment]
The safer thing to do would be to interpret a machine at a specific state that has that persisted actor ref. Would that work for your use-case?

**Reply by bhvngt** [reply]
Thanks @davidkpiano for your reply.

Here's my use case.

I have a top level state machine `R` that spawns couple of actors - say actor `A` and `B`. Now, actor `B` needs to get some information from actor `A` while transitioning to some state. This all works out great with the new `system` realm that allows me to stop passing around actors.

However, for the purpose of unit testing actor `B`, I was exploring a way to inject actor `A` into the `system` realm used by `B`. This will allow me to bypass instantiating top level state machine `R` in order to test `B`.

I tried system's _set method and it works out well. The leading underscore made me think that its probably intended as an internal method. Hence, was checking if it is safe or is there a better way to achieve my objective.

**Reply by davidkpiano** [reply, marked as answer]
I think for now, especially because XState v5 is still in beta, it's fine 👍

**Comment by bhvngt** [comment]
> I think for now, especially because XState v5 is still in beta, it's fine 👍

@davidkpiano Now that `v5` is released. Is there any recommended way to inject actors into the `system` for the purpose of unit testing the consuming actor.

I am able to inject and test them with unpublished api as follows

```typescript
import { type ActorRefFrom, createActor, fromPromise, waitFor } from "xstate";

const m1 = fromPromise(() => {
  return Promise.resolve(1);
});

const m2 = fromPromise(({ system }) => {
  const value = system.get("a1") as ActorRefFrom<typeof m1>;
  console.log(value.getSnapshot().status, value.getSnapshot().output); //?
  return Promise.resolve(2);
});

const a2 = createActor(m2);
// @ts-ignore
a2.system._set("a1", createActor(m1).start())
a2.start()
await waitFor(a2.system.get("a1"), (sa1) => {
  return sa1.output === 1
})
await waitFor(a2, (sa2) => {
  return sa2.output === 2
})
```

However, since `_set` is not published, typescript throws an error on that.
